src/box/lua/schema.lua

```diff
@@ -1619,6 +1620,42 @@ local function func_id_by_name(func_name, level)
 end
 box.internal.func_id_by_name = func_id_by_name -- for space.upgrade
 
+-- Normalize array of fields `fields`:
+-- - fields referenced as name are resolved to 0-based field number.
+-- - fields referenced as 1-based field number are converted to 0-based.
+local function normalize_fields(fields, format, name, level)
+    -- Do not much care if opts.covers is something strange like
+    -- sparse array or map with keys.
+    local result = {}
+    for i, field in ipairs(fields or {}) do
+        local idx
+        local field_ref = name .. "[" .. i .. "]: "
+        if type(field) == 'string' then
+            idx = format_field_index_by_name(format, field)
+            if idx == nil then
+                box.error(box.error.ILLEGAL_PARAMS, field_ref ..
+                          "field was not found by name '" .. field .. "'",
+                          level + 1)
+            end
+        elseif type(field) == 'number' then
+            if field <= 0 then
+                box.error(box.error.ILLEGAL_PARAMS, field_ref ..
+                          "field (number) must be one-based", level + 1)
+            end
+            idx = field
+        else
+            box.error(box.error.ILLEGAL_PARAMS, field_ref ..
+                      "field (name or number) is expected", level + 1)
+        end
+        result[i] = idx - 1
+    end
+    if #result ~= 0 then
```

Comment:
Better not call this function if `fields` is `nil` rather than handling this case here.